lints/type_cosplay/src/lib.rs

```diff
@@ -61,31 +64,63 @@ struct TypeCosplay {
     deser_types: FxHashMap<AdtKind, Vec<(DefId, Span)>>,
 }
 
+// get type X
+// check if implements Discriminator
+// check corresponding function call type:
+// if !try_deserialize
+// emit lint with warning to use try_deserialize (because any types deriving Discriminator should since it guaranteed to check discrim)
+
 impl<'tcx> LateLintPass<'tcx> for TypeCosplay {
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
         if_chain! {
             if !expr.span.from_expansion();
-            if let ExprKind::Call(fnc_expr, args_exprs) = expr.kind;
-            if is_deserialize_function(cx, fnc_expr);
+            if let ExprKind::Call(fnc_expr, _args_exprs) = expr.kind;
+            // TODO: is the following if statement really needed?? don't think it's ever used.
+            // all it does is check if AccountInfo.data is referenced...but what bytes we deser
+            // from shouldn't impact if this is a type-cosplay issue or not.
             // walk each argument expression and see if the data field is referenced
-            if args_exprs.iter().any(|arg| {
-                visit_expr_no_bodies(arg, |expr| contains_data_field_reference(cx, expr))
-            });
-            // get the type that the function was called on, ie X in X::deser()
+            // TODO: maybe just check arg is a byte array
+            // if args_exprs.iter().any(|arg| {
+            //     visit_expr_no_bodies(arg, |expr| contains_data_field_reference(cx, expr))
+            // });
+            // get the type that the function was called on, ie X in X::call()
             if let ExprKind::Path(qpath) = &fnc_expr.kind;
             if let QPath::TypeRelative(ty, _) = qpath;
             if let TyKind::Path(ty_qpath) = &ty.kind;
             let res = cx.typeck_results().qpath_res(ty_qpath, ty.hir_id);
             if let Res::Def(_, def_id) = res;
+            let middle_ty = cx.tcx.type_of(def_id);
+            if let Some(trait_did) = get_trait_def_id(cx, &paths::ANCHOR_DISCRIMINATOR_TRAIT);
```

Comment:
I think this will return false if this trait is not defined.

Reply:
mm right that's bad. I think if I moved line 93 into the first if statement then it would be ok. if the trait is not defined then, it would just jump to the else block and execute the good ol check.